machinelearning/tensorflow.md

```diff
@@ -0,0 +1,349 @@
+  OMP_PROC_BIND = 'FALSE'
+  OMP_PLACES = ''
+  OMP_STACKSIZE = '0'
+  OMP_WAIT_POLICY = 'PASSIVE'
+  OMP_THREAD_LIMIT = '4294967295'
+  OMP_MAX_ACTIVE_LEVELS = '1'
+  OMP_CANCELLATION = 'FALSE'
+  OMP_DEFAULT_DEVICE = '0'
+  OMP_MAX_TASK_PRIORITY = '0'
+  OMP_DISPLAY_AFFINITY = 'FALSE'
+  OMP_AFFINITY_FORMAT = 'level %L thread %i affinity %A'
+  OMP_ALLOCATOR = 'omp_default_mem_alloc'
+  OMP_TARGET_OFFLOAD = 'DEFAULT'
+  GOMP_CPU_AFFINITY = ''
+  GOMP_STACKSIZE = '0'
+  GOMP_SPINCOUNT = '300000'
+OPENMP DISPLAY ENVIRONMENT END
+
+```
+4. The above triaging steps cover the performance issues due to the missing compile time, runtime or library level optimizations across the Tensorflow stack. If the troubleshooting is inconclusive or stuck with any step mentioned above, please raise a support request on [aws-graviton-getting-started](https://github.com/aws/aws-graviton-getting-started) github.
+
+
+# Building Tensorflow from sources
+
+The above mentioned DLAMI/DLC/python wheels provide the stable release binaries with the optimzations that are part of upstream repositories. If someone is interested in working with the master branch or incorporate the downstream optimizations, please follow the below steps to build tensorflow from sources, and follow the runtime configuration steps, provided above, to enable onednn+acl backend for optimal performance. 
+
+```
+# install bazel for aarch64
+mkdir bazel
+cd bazel
+wget https://github.com/bazelbuild/bazel/releases/download/5.1.1/bazel-5.1.1-linux-arm64
+mv bazel-5.1.1-linux-arm64 bazel
+chmod a+x bazel
+export PATH=/home/ubuntu/bazel/:$PATH
+
+#clone tensorflow repository
+git clone https://github.com/tensorflow/tensorflow.git
+cd tensorflow
+git checkout <latest stable version>
+
+#set the build configuration
+export HOST_C_COMPILER=(which gcc)
+export HOST_CXX_COMPILER=(which g++)
+export PYTHON_BIN_PATH=(which python)
+export USE_DEFAULT_PYTHON_LIB_PATH=1
+export TF_ENABLE_XLA=1
+export TF_DOWNLOAD_CLANG=0
+export TF_SET_ANDROID_WORKSPACE=0
+export TF_NEED_MPI=0
+export TF_NEED_ROCM=0
+export TF_NEED_GCP=0
+export TF_NEED_S3=0
+export TF_NEED_OPENCL_SYCL=0
+export TF_NEED_CUDA=0
+export TF_NEED_HDFS=0
+export TF_NEED_OPENCL=0
+export TF_NEED_JEMALLOC=1
+export TF_NEED_VERBS=0
+export TF_NEED_AWS=0
+export TF_NEED_GDR=0
+export TF_NEED_OPENCL_SYCL=0
+export TF_NEED_COMPUTECPP=0
+export TF_NEED_KAFKA=0
+export TF_NEED_TENSORRT=0
+
+./configure
+
+#issue bazel build command with 'mkl_aarch64' config to enable onednn+acl backend
+bazel build --verbose_failure -s --config=mkl_aarch64  //tensorflow/tools/pip_package:build_pip_package //tensorflow:libtensorflow_cc.so //tensorflow:install_headers
+
+#Create and install the wheel
+./bazel-bin/tensorflow/tools/pip_package/build_pip_package ./wheel-TF2.9.0-py3.8-aarch64
+
+#Output wheel file is in: /home/ubuntu/tensorflow/wheel-TF2.9.0-py3.8-aarch64
+pip install <wheel-TF2.9.0-py3.8-aarch64/*.whl>
+
+```
+
+# Enabling XLA optimizations
```

Comment:
do you want this after the building from source?

Reply:
shouldn't be be close to the grapper discussion?

Reply:
My thought process was to give all the details about how to use the existing packages, thats where Grappler comes as it is enabled by default, and then talk about how to build from sources, and then introduce how to build the downstream features with PRs. Once the PRs get merged, these sections taken out of source builds, and move them up.